src/pymor/reductors/neural_network.py

```diff
@@ -296,22 +298,119 @@ def build_basis(self):
+                mu_tensor = torch.DoubleTensor(mu.to_numpy())
+                u_tensor = torch.DoubleTensor(reduced_basis.inner(u)[:,0])
+                samples.append((mu_tensor, u_tensor))
+            return samples
+
         def reconstruct(self, u):
             """Reconstruct high-dimensional vector from reduced vector `u`."""
             assert hasattr(self, 'reduced_basis')
             return self.reduced_basis.lincomb(u.to_numpy())
 
 
+    class NeuralNetworkInstationaryReductor(NeuralNetworkReductor):
+        """Reduced Basis reductor for instationary problems relying on
+        artificial neural networks.
+
+        This is a reductor that constructs a reduced basis using proper
+        orthogonal decomposition and trains a neural network that approximates
+        the mapping from parameter and time space to coefficients of the
+        full-order solution in the reduced basis.
+        The approach is described in [WHR19]_.
+
+        Parameters
+        ----------
+        fom
+            The full-order |Model| to reduce.
+        training_set
+            Set of |parameter values| to use for POD and training of the
+            neural network.
+        validation_set
+            Set of |parameter values| to use for validation in the training
+            of the neural network.
+        validation_ratio
+            Fraction of the training set to use for validation in the training
+            of the neural network (only used if no validation set is provided).
+        basis_size
+            Desired size of the reduced basis. If `None`, rtol, atol or l2_err must
+            be provided.
+        rtol
+            Relative tolerance the basis should guarantee on the training set.
+        atol
+            Absolute tolerance the basis should guarantee on the training set.
+        l2_err
+            L2-approximation error the basis should not exceed on the training
+            set.
+        pod_params
+            Dict of additional parameters for the POD-method.
+        ann_mse
+            If `'like_basis'`, the mean squared error of the neural network on
+            the training set should not exceed the error of projecting onto the basis.
+            If `None`, the neural network with smallest validation error is
+            used to build the ROM.
+            If a tolerance is prescribed, the mean squared error of the neural
+            network on the training set should not exceed this threshold.
+            Training is interrupted if a neural network that undercuts the
+            error tolerance is found.
+        """
+
+        def __init__(self, fom, training_set, validation_set=None, validation_ratio=0.1,
+                     basis_size=None, rtol=0., atol=0., l2_err=0., pod_params=None,
+                     ann_mse='like_basis'):
+            assert 0 < validation_ratio < 1 or validation_set
+            self.__auto_init(locals())
+            self.nt = fom.time_stepper.nt
```

Comment:
I think we should not assume that every `fom` has a `time_stepper`. Many models coming from an external solver won't. How about deriving `nt` from the length of the array returned by `fom.solve` and asserting that all have the same length?

Reply:
Sounds good, I will try that.

Reply:
`fom.solve` is never called in the functions overridden by `NeuralNetworkInstationaryReductor`, it is only called in the functions inherited by `NeuralNetworkReductor`. Hence, I am not sure where to put the computation of `nt`.

Reply:
If `_compute_samples` is only called for a single snapshot, you can determine `nt` in there.